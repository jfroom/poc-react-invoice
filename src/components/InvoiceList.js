import React from 'react';
import {Button, Table, Glyphicon} from 'react-bootstrap';
import {LinkContainer} from 'react-router-bootstrap'
import PropTypes from 'prop-types'
import NumberFormat from 'react-number-format'

const InvoiceList = ({ invoices, onClickDelete }) => (
  <div>
    <h1>Invoices</h1>
    <hr/>

    <div className='col-sm-12 text-right'>
      <LinkContainer to="/new">
        <Button bsStyle='success' bsSize='large'><Glyphicon glyph='plus'/> New Invoice</Button>
      </LinkContainer>
    </div>
    <p/>

    <Table condensed hover className='invoice-list'>
      <thead>
        <tr>
          <th>Title</th>
          <th>Total</th>
          <th>Status</th>
          <th>Action</th>
        </tr>
      </thead>
      <tbody>
        {invoices.map(invoice =>
          <tr key={invoice.id}>
            <td>{invoice.title}</td>
            <td><NumberFormat value={invoice.total} displayType={'text'} thousandSeparator={true} decimalPrecision={true} prefix={'$'} /></td>
            <td>{invoice.status}</td>
            <td>
              <Button className='btn-xs btn-info'><Glyphicon glyph='envelope'/></Button>
              <LinkContainer to={"/edit/" + invoice.id} className='btn-xs btn-primary'>
                <Button><Glyphicon glyph='pencil'/></Button>
              </LinkContainer>
              <Button className='btn-xs btn-danger' onClick={() => onClickDelete(invoice.id)}><Glyphicon glyph='trash'/></Button>
            </td>
          </tr>
        )}
      </tbody>
    </Table>
  </div>
)

InvoiceList.propTypes = {
  invoices: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
    total: PropTypes.number.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired
    }).isRequired).isRequired,
    notes: PropTypes.text
  }).isRequired).isRequired
}

export default InvoiceList
