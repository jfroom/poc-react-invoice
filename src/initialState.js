const initialState =
{
  invoices: {
    nextId: 3,
    invoices: [
      {
        id: 1,
        title: "Hoyt Residence Kitchen Remodel",
        status: "Unpaid",
        total: 7000,
        items: [
          {
            text: 'Quartz Counters',
            price: 5000
          },
          {
            text: 'Subway Tile Backsplash',
            price: 2000
          }
        ],
        notes: "Thank you for your business!"
      },
      {
        id: 2,
        title: "Lovejoy Residence Bathroom Remodel",
        status: "Paid",
        total: 7000.99,
        items: [
          {
            text: 'Soaker tub',
            price: 1000.99
          },
          {
            text: 'Tile work',
            price: 2000
          }
        ],
        notes: "Payment due net-30."
      }
    ]
  }
}
export default initialState
