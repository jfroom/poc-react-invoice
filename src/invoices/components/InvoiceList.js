import React, {Component} from 'react';
import {Button, Table, Glyphicon} from 'react-bootstrap';
import {LinkContainer} from 'react-router-bootstrap'
import PropTypes from 'prop-types'
import NumberFormat from 'react-number-format'

class InvoiceList extends Component {
  render() {
    const { invoices, handleDeleteInvoice } = this.props

    const getMailToLink = (invoice) => {
      const subject = encodeURI("Invoice: " + invoice.title)
      const body = "Please view your invoice at this web page: http://example.com/invoices/" + invoice.id
      return "mailto:?subject=" + subject + "&body=" + body
    }
    return (
      <div>
        <h1>Invoices</h1>
        <hr/>

        <div className='col-sm-12 text-right'>
          <LinkContainer to="/new">
            <Button bsStyle='success' bsSize='large'><Glyphicon glyph='plus'/> New Invoice</Button>
          </LinkContainer>
        </div>
        <p/>

        <Table condensed hover className='invoice-list'>
          <thead>
            <tr>
              <th>Title</th>
              <th>Total</th>
              <th>Status</th>
              <th>Action</th>
            </tr>
          </thead>
          <tbody>
            {invoices.map(invoice =>
              <tr key={invoice.id}>
                <td>{invoice.title}</td>
                <td><NumberFormat value={invoice.total} displayType={'text'} thousandSeparator={true} decimalPrecision={true} prefix={'$'} /></td>
                <td>{invoice.status}</td>
                <td>
                  <a target='blank' href={getMailToLink(invoice)}><Button className='btn-xs btn-info'><Glyphicon glyph='envelope'/></Button></a>
                  <LinkContainer to={"/edit/" + invoice.id} className='btn-xs btn-primary'>
                    <Button><Glyphicon glyph='pencil'/></Button>
                  </LinkContainer>
                  <Button className='btn-xs btn-danger' onClick={() => handleDeleteInvoice(invoice.id)}><Glyphicon glyph='trash'/></Button>
                </td>
              </tr>
            )}
          </tbody>
        </Table>
      </div>
    )
  }
}

InvoiceList.propTypes = {
  invoices: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
    total: PropTypes.number.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      price: PropTypes.number.isRequired
    }).isRequired).isRequired,
    notes: PropTypes.text
  }).isRequired).isRequired,
  handleDeleteInvoice: PropTypes.func.isRequired
}

export default InvoiceList
