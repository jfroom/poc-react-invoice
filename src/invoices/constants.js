// @flow
export const StatusTypes = {
  NONE: '',
  PAID: 'Paid',
  UNPAID: 'Unpaid',
}
