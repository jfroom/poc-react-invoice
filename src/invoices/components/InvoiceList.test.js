import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import InvoiceList from './InvoiceList'
import initialState from '../../initialState'
import { StatusTypes } from '../constants'

let props = {}
let handleDeleteInvoice
let handleChangeStatusFilter

beforeEach( () => {
  handleDeleteInvoice = jest.fn()
  handleChangeStatusFilter = jest.fn()
  props = { handleDeleteInvoice, handleChangeStatusFilter}
})

it('empty', () => {
  const wrap = shallow(<InvoiceList {...props} />)
  expect(toJson(wrap)).toMatchSnapshot()

  // Empty table
  expect(wrap.find('.invoice-list tbody tr')).toHaveLength(0)
})

it('full list with delete & status filter mocks', () => {
  props.invoices = initialState.invoices.invoices

  const wrap = shallow(<InvoiceList {...props} />)
  expect(toJson(wrap)).toMatchSnapshot()

  // Populated table
  expect(wrap.find('.invoice-list tbody tr')).toHaveLength(3)

  // Delete callback
  wrap.find(".invoice-list tbody tr .btn-delete").at(0).simulate('click')
  expect(handleDeleteInvoice.mock.calls).toEqual([[1]])

  // Change filter
  wrap.find("#status-filter").first().simulate('change', {target: {value: StatusTypes.Paid}})
  expect(handleChangeStatusFilter.mock.calls).toEqual([[StatusTypes.Paid]])
})
