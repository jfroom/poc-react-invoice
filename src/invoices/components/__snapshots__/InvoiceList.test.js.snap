// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty 1`] = `
<styled.div>
  <h1>
    Invoices
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-9"
    >
      <LinkContainer
        activeClassName="active"
        exact={false}
        replace={false}
        strict={false}
        to="/new"
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="large"
          bsStyle="success"
          disabled={false}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="plus"
          />
           
           New Invoice
        </Button>
      </LinkContainer>
    </div>
    <div
      className="col-sm-3"
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
        >
          Status Filter
        </label>
        <select
          className="form-control"
          id="status-filter"
          onChange={[Function]}
          value="Paid"
        >
          <option
            value=""
          >
            View All
          </option>
          <option
            value="Unpaid"
          >
            Unpaid
          </option>
          <option
            value="Paid"
          >
            Paid
          </option>
        </select>
      </div>
    </div>
  </div>
  <Table
    bordered={false}
    bsClass="table"
    className="invoice-list well"
    condensed={true}
    hover={true}
    responsive={false}
    striped={false}
  >
    <thead>
      <tr>
        <th>
          Due Date
        </th>
        <th>
          Title
        </th>
        <th>
          Total
        </th>
        <th>
          Status
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody />
  </Table>
  <FormGroup
    bsClass="form-group"
    controlId="total"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Grand Total
    </ControlLabel>
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      $0.00
    </FormControlStatic>
  </FormGroup>
</styled.div>
`;

exports[`full list with delete & status filter mocks 1`] = `
<styled.div>
  <h1>
    Invoices
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-9"
    >
      <LinkContainer
        activeClassName="active"
        exact={false}
        replace={false}
        strict={false}
        to="/new"
      >
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsSize="large"
          bsStyle="success"
          disabled={false}
        >
          <Glyphicon
            bsClass="glyphicon"
            glyph="plus"
          />
           
           New Invoice
        </Button>
      </LinkContainer>
    </div>
    <div
      className="col-sm-3"
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
        >
          Status Filter
        </label>
        <select
          className="form-control"
          id="status-filter"
          onChange={[Function]}
          value="Paid"
        >
          <option
            value=""
          >
            View All
          </option>
          <option
            value="Unpaid"
          >
            Unpaid
          </option>
          <option
            value="Paid"
          >
            Paid
          </option>
        </select>
      </div>
    </div>
  </div>
  <Table
    bordered={false}
    bsClass="table"
    className="invoice-list well"
    condensed={true}
    hover={true}
    responsive={false}
    striped={false}
  >
    <thead>
      <tr>
        <th>
          Due Date
        </th>
        <th>
          Title
        </th>
        <th>
          Total
        </th>
        <th>
          Status
        </th>
        <th>
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
        onClick={[Function]}
      >
        <td>
          1/1/2020
        </td>
        <td>
          Hoyt Residence Kitchen Remodel
        </td>
        <td>
          $7,000.00
        </td>
        <td
          className="status"
        >
          Unpaid
        </td>
        <td>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-primary"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="pencil"
            />
          </Button>
          <a
            href="mailto:?subject=Invoice:%20Hoyt%20Residence%20Kitchen%20Remodel&body=Please view your invoice at this web page: http://example.com/invoices/1"
            onClick={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-xs btn-info"
              disabled={false}
            >
              <Glyphicon
                bsClass="glyphicon"
                glyph="envelope"
              />
            </Button>
          </a>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-danger btn-delete"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="trash"
            />
          </Button>
        </td>
      </tr>
      <tr
        className=""
        onClick={[Function]}
      >
        <td>
          5/1/2017
        </td>
        <td>
          Lovejoy Residence Bathroom Remodel
        </td>
        <td>
          $3,000.99
        </td>
        <td
          className="status"
        >
          Paid
        </td>
        <td>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-primary"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="pencil"
            />
          </Button>
          <a
            href="mailto:?subject=Invoice:%20Lovejoy%20Residence%20Bathroom%20Remodel&body=Please view your invoice at this web page: http://example.com/invoices/2"
            onClick={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-xs btn-info"
              disabled={false}
            >
              <Glyphicon
                bsClass="glyphicon"
                glyph="envelope"
              />
            </Button>
          </a>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-danger btn-delete"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="trash"
            />
          </Button>
        </td>
      </tr>
      <tr
        className="overdue"
        onClick={[Function]}
      >
        <td>
          5/10/2017
        </td>
        <td>
          Glisan Residence Living Room Remodel
        </td>
        <td>
          $3,090.99
        </td>
        <td
          className="status"
        >
          Unpaid
        </td>
        <td>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-primary"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="pencil"
            />
          </Button>
          <a
            href="mailto:?subject=Invoice:%20Glisan%20Residence%20Living%20Room%20Remodel&body=Please view your invoice at this web page: http://example.com/invoices/3"
            onClick={[Function]}
            rel="noopener noreferrer"
            target="_blank"
          >
            <Button
              active={false}
              block={false}
              bsClass="btn"
              bsStyle="default"
              className="btn-xs btn-info"
              disabled={false}
            >
              <Glyphicon
                bsClass="glyphicon"
                glyph="envelope"
              />
            </Button>
          </a>
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-xs btn-danger btn-delete"
            disabled={false}
            onClick={[Function]}
          >
            <Glyphicon
              bsClass="glyphicon"
              glyph="trash"
            />
          </Button>
        </td>
      </tr>
    </tbody>
  </Table>
  <FormGroup
    bsClass="form-group"
    controlId="total"
  >
    <ControlLabel
      bsClass="control-label"
      srOnly={false}
    >
      Grand Total
    </ControlLabel>
    <FormControlStatic
      bsClass="form-control-static"
      componentClass="p"
    >
      $13,091.98
    </FormControlStatic>
  </FormGroup>
</styled.div>
`;
